[CoSim] Generic CoSimIO

philbucher opened the pull request:
This is a first working version for the generic CoSimIO. The basics are working but minor things have to be added, as well as cleanup and tests. Also this branch contains several fixes that I will port in smaller PRs

It currently supports only IO based on Files but is intended to be extended to other communication formats such as Sockets or MPI. The selection of the communication format is done at runtime.

I added an example dummy solver that is pure cpp, but still provides the "full interface" for synchronization, not only synchronization based on files. 
All coupling options are shown, [weak](https://github.com/KratosMultiphysics/Kratos/blob/cosim/cosim_io/applications/CoSimulationApplication/tests/dummy_solver/cpp/dummy_solver.cpp#L147-L173) & [strong](https://github.com/KratosMultiphysics/Kratos/blob/cosim/cosim_io/applications/CoSimulationApplication/tests/dummy_solver/cpp/dummy_solver.cpp#L175-L209) coupling based on using the file-comm for synchronization as well as the [full interface](https://github.com/KratosMultiphysics/Kratos/blob/cosim/cosim_io/applications/CoSimulationApplication/tests/dummy_solver/cpp/dummy_solver.cpp#L211-L268)

Exposure of the functionalities to fortran is planned, but I have not had time yet to do it. The mechanism will be the same as what was used for Empire, hence we don't have to come up with a new solution, but rather implement an existing and working solution.
Besides I think the dummy cpp solver is reduced to the minimum such that it should be understandable what is happening.

part of #5140

philbucher committed:
adding PrintTiming member

philbucher committed:
specifying identifier

philbucher committed:
more advances in file-comm

philbucher committed:
fixing sending of control_signal

philbucher committed:
minor

philbucher committed:
adding IsConnectionMaster

philbucher committed:
improve member visubility

philbucher committed:
towards comm in folder

philbucher committed:
leftover

philbucher committed:
better folder name

philbucher committed:
moving auto-disconnect to avoid pure virtual fct call

philbucher committed:
missing

philbucher committed:
correting name

philbucher committed:
almost finished with mesh im-/export

philbucher committed:
improvements of dummy-solver

philbucher committed:
improving cosim-params

philbucher committed:
improving solver-wrapper

philbucher committed:
adding conv-signal output

philbucher committed:
towards improving Send/Recv control signal

philbucher committed:
adding interface for controlsignal

philbucher committed:
implemented send/recv control signal for FileComm

philbucher committed:
sending control signal has to be blocking
We could implement some sort of queue as a remedy

philbucher committed:
correcting default echo-lvl

philbucher committed:
using name of interface-data as identifier

Solver name is automatically added

philbucher committed:
same in wrapper

philbucher committed:
making full-control work

philbucher committed:
fixing cosim-params

philbucher committed:
minor fix

philbucher committed:
cleanup

philbucher committed:
adding todo

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_api_tools.h:
```diff
@@ -0,0 +1,60 @@
+#include <unordered_map>
+
+namespace CoSim {
+namespace Tools {
```
Internals?

philbucher replied:
done

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_api_tools.h:
```diff
@@ -0,0 +1,60 @@
+#define CS_LOG std::cout << "[CoSimIO] "
+#define CS_LOG_IF(condition) if(condition) CS_LOG
+
+typedef std::unordered_map<std::string, std::string> SettingsType;
```
A JSON string wouldn't be better?

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_api_tools.h:
```diff
@@ -0,0 +1,60 @@
+namespace CoSim {
+namespace Tools {
+
+#define CS_LOG std::cout << "[CoSimIO] "
```
I think we should add a KRATOS_INDEPENDENT macro and change all these details depending on if someone is strictly avoiding the link to Kratos or not.

pooyan-dadvand replied:
I would also check if the macro is not defined. This would let the users use their on logger if available and for us to substitute it for Fortran.

philbucher replied:
I improved this, but I am not sure what you want can be achieved if we split header/source

The macros will be used inside the source-file, and this gets precompiled I assume before fortran links against it. I.e. by the time fortran replaces those macros most of them are already compiled

Probably better to discuss in person

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_api_tools.h:
```diff
@@ -0,0 +1,60 @@
+    SettingsType settings;
+    std::string current_line;
+    while (std::getline(settings_file, current_line)) {
+        // TODO implement this
```
I understand that using JSON would need an additional dependency here... but it unifies the configure in all places.

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_comm.h:
```diff
@@ -0,0 +1,155 @@
+
+    std::string GetName() const        {return mName;}
+    int GetEchoLevel() const           {return mEchoLevel;}
+    bool GetIsConnectionMaster() const {return mIsConnectionMaster;}
```
?

philbucher replied:
one side (does not matter which one) should be the master to perform some operations.
e.g. for File-Communication this would be the creation / deletion of the folder in which the comm happens
for MPI-Comm one side would open ports, the other side would connect to it
=> the different sides do different calls

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_comm.h:
```diff
@@ -0,0 +1,155 @@
+    bool mPrintTiming = false;
+    bool mIsConnected = false;
+
+    virtual bool ConnectDetail() = 0;
```
private pure virtual function? Shouldn't be = delete?

philbucher replied:
this is supposed to be a private interface. The derived class has to implement it but cannot call it

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_data_containers.h:
```diff
@@ -0,0 +1,45 @@
+        std::vector<int>& cell_types;
+    };
+
+    struct Data
```
Maybe in the C++ version, we can have this with a template of data type and then register/expose it with typical data types.

philbucher replied:
ok sounds good

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_file_comm.h:
```diff
@@ -0,0 +1,419 @@
+        return true;
+    }
+
+    void SendControlSignalDetail(const int rSignal, const std::string& rIdentifier) override
```
I know that this is an implementation detail, but I would use the same method names used in data_communicator for coherence.

philbucher replied:
hm not sure what you mean
This has nothing to do with MPI

pooyan-dadvand commented on applications/CoSimulationApplication/co_simulation_api/co_sim_io_version.h:
```diff
@@ -0,0 +1,23 @@
+#define KRATOS_CO_SIM_IO_VERSION_H_INCLUDED
+
+constexpr int GetMajorVersion() {
+    return 1;
```
I would use the same versioning as Kratos to have a clear figure of compatibility.

philbucher replied:
this is not used yet, I just added it as a reminder

pooyan-dadvand commented on applications/CoSimulationApplication/python_scripts/co_simulation_tools.py:
```diff
@@ -27,6 +27,29 @@ def UsingPyKratos():
     return False
 
 
+control_signal_map = {
```
This map shouldn't be in cpp? Or it is to be defined by users?

philbucher replied:
solved/improved

pooyan-dadvand commented on applications/CoSimulationApplication/python_scripts/solver_wrappers/dummy_solver/dummy_solver_io.py:
```diff
@@ -0,0 +1,93 @@
+        pass
+
+
+def ParametersToStringDict(param):
```
Why we need this?

pooyan-dadvand commented on applications/CoSimulationApplication/python_scripts/solver_wrappers/dummy_solver/dummy_solver_io.py:
```diff
@@ -0,0 +1,93 @@
+
+    return string_dict
+
+def GetDataLocation(location_str):
```
And this?

philbucher replied:
this specifies where on the modelpart the data is. e.g. node, element, node-nonhist, condition, modelpart

pooyan-dadvand commented on applications/CoSimulationApplication/tests/dummy_solver/cpp/dummy_solver.cpp:
```diff
@@ -0,0 +1,327 @@
+        // receive control signal to decide what to do
+        // the signals are defined in KratosMultiphysics/applications/CoSimulationApplication/python_scripts/co_simulation_tools.py
+        control_signal = co_sim_io.RecvControlSignal(identifier);
+        if (control_signal == 1) {
```
I assume that these control signals should be given with some constexp or enums.

philbucher replied:
solved

pooyan-dadvand reviewed (commented):
@philbucher, Thanks for the fast prototyping. Very much aligned with what we discussed. 

However, seems that the freedom to compile without Kratos gives huge implementation overhead. 

Maybe we can compile only against the KratosCore (not Kratos) or we may define a base layer in the core (including basic tools which are the base dependencies in kratos like exception, logger, parameter) and make them KRATOS_INDEPENDENT to be used here to have more coherence and less duplicated code.

philbucher committed:
Merge branch 'master' into cosim/cosim_io

philbucher committed:
adding co_sim_io_define

philbucher committed:
including co_sim_io_define

philbucher committed:
overridding the macro definitions
Using the Kratos macros for the CoSimIO inside of Kratos

philbucher committed:
Tools => Internals

philbucher committed:
adding enum for ControlSignals

philbucher committed:
exposing control signals

philbucher committed:
using control signals from C++

philbucher committed:
minor

philbucher committed:
using the control signals

philbucher committed:
adding Kratos:: - TBD

philbucher committed:
improving co_sim_define
Now the macros can be redefined

philbucher committed:
using new macros

philbucher committed:
implemented Function registration

philbucher committed:
implemented AdvanceInTime

philbucher committed:
Also adding a header/source version of the IO
To be discussed which one to use

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher commented on applications/CoSimulationApplication/co_simulation_api/co_sim_io_define.h:
```diff
@@ -0,0 +1,48 @@
+//  Main authors:    Philipp Bucher
+//
+
+#ifndef KRATOS_CO_SIM_IO_DEFINE_H_INCLUDED
```
@pooyan-dadvand in this file I check if the macros have been defined yet. If not I use default ones, otherwise they can be defined from outside

pooyan-dadvand replied:
:+1:

philbucher commented on applications/CoSimulationApplication/co_simulation_api/co_sim_io_impl.h:
```diff
@@ -0,0 +1,229 @@
+}
+
+
+inline void CoSimIO::Run()
```
now I managed to completely hide this inside the API
The client only has to register the functions

philbucher commented on applications/CoSimulationApplication/custom_python/add_co_sim_io_to_python.cpp:
```diff
@@ -0,0 +1,372 @@
+#define KRATOS_CO_SIM_ERROR KRATOS_ERROR
+#define KRATOS_CO_SIM_ERROR_IF KRATOS_ERROR_IF
+#define KRATOS_CO_SIM_ERROR_IF_NOT KRATOS_ERROR_IF_NOT
+#include "co_simulation_api/co_sim_io.h"
```
Here I define the macros before including the header. This way I can use the macros of the including code
Not sure if this is what you want / intended @pooyan-dadvand

pooyan-dadvand replied:
I was thinking a little bit different but this also does the same!

philbucher commented on applications/CoSimulationApplication/tests/dummy_solver/cpp/dummy_solver.cpp:
```diff
@@ -0,0 +1,336 @@
+    co_sim_io.Disconnect();
+}
+
+void RunSolutionCoSimulationOrchestrated(MeshType& rMesh, DataFieldType& rDataField)
```
this function is the full integration of CoSim into an external solver

(ignore the `new`/`delete`, this is a temp hack)

philbucher commented on kratos/includes/exception.h:
```diff
@@ -152,11 +152,11 @@ namespace Kratos
   ///@name Kratos Macros
   ///@{
 
-#define KRATOS_ERROR throw Exception("Error: ", KRATOS_CODE_LOCATION)
```
sorry for the changes in the core
We will discuss this together with the macro issue

philbucher reviewed (commented):
I did major updates, now it is in line with what we discussed @pooyan-dadvand 
- I have not yet implemented the interface for fortran (no time), but with the tests I did it should be relatively straight forward to do it
- Regarding the linking to the core: Now the only thing where we would use the Core for is the `Parameters`. IMO this is NOT worth it, especially because the "settings" I need here are very minimal. I only need sth very very basic. I am quite sure it would cause more problems that it solves.
- I created a header-only and a header/source version of the API. I am not sure which one is better. Even though I like header-only, also having the header/source version has the significant advantage that we could use the SAME interface for C & C++ (using `extern C`). This means one interface less to maintain, and then we can also directly use it in fortran.
- One open question is how to nicely handle distributed data, i.e. MPI-runs. As of now there is no special mechanism for this, we can think abt this.

philbucher committed:
simplified data-container registration

philbucher committed:
simplified function registration

philbucher committed:
using header-only version again

philbucher committed:
adding check

pooyan-dadvand commented:
Very nice improvement!

philbucher commented:
> Very nice improvement!

thx, but it is still a lot WIP
We should discuss how to continue ... :)

pooyan-dadvand commented:
Just let me know when is convenient for you

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher committed:
renamed folder

philbucher committed:
removing non-header-only version again

philbucher committed:
fix paths

philbucher committed:
missing rename

philbucher committed:
move implementation to subfolder

philbucher committed:
moving large part of IO
towards unique interface across languages

philbucher committed:
put more fcts in interface

philbucher committed:
started working on C-Interface

philbucher committed:
dummy for fortran interface

philbucher committed:
moving fct to private

philbucher committed:
init version of dummy c solver

philbucher committed:
moving more functionalities to new interface

philbucher committed:
more fct-ptr

philbucher committed:
moving helpers to separate folder and adding C-Lib

philbucher committed:
minor

philbucher committed:
adding empty readme for dummy-solvers

philbucher committed:
starting using C-IO

philbucher committed:
include IO in Fortran dummy solver

philbucher committed:
link fortran dummy against C-IO

philbucher committed:
Merge branch 'master' into cosim/cosim_io

philbucher committed:
adding module for fortran - IS THIS CORRECT???

philbucher committed:
fixing throwing in destructor

philbucher committed:
removed data_containers
only compiling, not running

philbucher committed:
more code commented for updates

philbucher committed:
updates class-name

philbucher committed:
update namespace name

philbucher committed:
leftover removed

philbucher committed:
update argument names

philbucher committed:
updating name arg

philbucher committed:
replaced old macros

philbucher committed:
minor

philbucher committed:
started with new mem-alloc

philbucher committed:
update usage of IsConnectionMaster

philbucher committed:
adding runfile for cpp-tests

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher committed:
adding wrappers for containers with test

philbucher committed:
cleaned interface of IO

philbucher committed:
adapted cpp dummy solver to changes in interface

philbucher committed:
worked on the C-Inteface

philbucher committed:
minor in CoSimComm

philbucher committed:
adpated implementation to changes in IO

philbucher committed:
starting to fill C-dummy solver

philbucher committed:
simplified interface to use refs instead of ptrs

philbucher committed:
adding resize_if_smaller

philbucher committed:
some cleanup

philbucher committed:
using perfect forwarding to simplify

philbucher committed:
WIP function registration

philbucher committed:
temp commented function

philbucher committed:
advances in Registration of functions
API more or less fixed for C++ and C

philbucher committed:
minor cleanup

philbucher committed:
restructured files for clarity I

philbucher committed:
more restructuring for clarity

philbucher committed:
renamed folder

philbucher committed:
moving impl to subfolder

philbucher committed:
removing unused version

philbucher committed:
adding FreeMemory for Fortran

philbucher committed:
Continued with Fortran interface

philbucher committed:
WIP in fortran dummy solver

philbucher committed:
added ImportMesh in F-interface

philbucher committed:
added missing functions to Fortran Interface

philbucher committed:
remaining methods added in F-Interface

philbucher committed:
cleanup in fortran dummy solver

philbucher committed:
changed all interface fcts to void
for better compatibility with fortran

philbucher committed:
copy-paste bugfix

philbucher committed:
correcting strong coupling of cpp-solver

philbucher committed:
improving readability

philbucher committed:
using macro for print

philbucher committed:
fixing conditional

philbucher committed:
major improvement in the dummy fortran solver

philbucher committed:
removing debug print

philbucher committed:
improved printing in fortran solver

philbucher committed:
bugfix function name

philbucher committed:
soecifying intent in fortran IO

philbucher committed:
improvements and cleanup in fortran solver

philbucher committed:
created functions to allocate Memory for Fortran in C
=> needed to manage memory in one language

philbucher committed:
bugfix

philbucher committed:
thoughts on async comm for sockets/mpi

philbucher committed:
started bringing back file-comm

philbucher committed:
passing sizes

philbucher committed:
using pass by val in fortran

philbucher committed:
minor renaming for consistency

philbucher committed:
better use of internals namespace

philbucher committed:
started fixing file-comm

philbucher committed:
initial tests for file-comm

philbucher committed:
minor bugfixes in file-comm

philbucher committed:
making info prints depending on echo-level

philbucher committed:
consistent renamings

philbucher committed:
reduce wait time
TODO maybe this should be configureable!

philbucher committed:
fully switching to C-style dyn-mem alloc

philbucher committed:
adding file-comm tests with raw mem

philbucher committed:
Merge branch 'master' into cosim/cosim_io

philbucher committed:
adding missing Connect-fct

philbucher committed:
fixing integration of Control-Signal exchang

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher committed:
minor correction passing conv-signal

philbucher committed:
update list initialization

philbucher committed:
adding test for file-comm of ControlSignal

philbucher committed:
fixing file-comm when connectivities don't start from 0

philbucher committed:
adding info print

philbucher committed:
adding tests for file-comm of Mesh

philbucher committed:
updates for file-comm of mesh

philbucher committed:
started bringing back py-exposure of IO

philbucher committed:
bugfix in test

philbucher committed:
adding comment related to folder creation

philbucher committed:
exposure almost ready

philbucher committed:
update dummy solver-IO

philbucher committed:
simplified Interface

philbucher committed:
minor synstax fix in cpp dummy solver

philbucher committed:
improved DataContainer

philbucher committed:
exposing ExportMesh

philbucher committed:
adding cpp-tests to CoSim-testscript
same style as in other apps

philbucher committed:
consistent renamings

philbucher committed:
brought back import/export of values to python

philbucher committed:
update kratos_cosim_io

philbucher committed:
only expose Connect with settings

philbucher committed:
more tests for DataContainers

philbucher committed:
refactor mok-fsi to separate file

philbucher committed:
adding ParametersToStringDict to cs_tools

philbucher committed:
fixing bug in case of no elements and add test

philbucher committed:
bugfix communication state of convergence

philbucher committed:
bugfixes py-exposure

philbucher committed:
corrections in KratosCoSimIO

philbucher committed:
adding function to cs-tools

philbucher committed:
adding method to co-sim-test-case to run with external

philbucher committed:
add check for return code

philbucher committed:
test for IO: Mok-FSI => structure external

philbucher committed:
adding Mok external structure test to validation

philbucher committed:
muting many prints

philbucher committed:
started adding tests for py-exposure

philbucher committed:
updates in dummy cpp solver

philbucher committed:
bugfix py-exposure

philbucher committed:
started test for dummy-sover and CoSimIO

philbucher committed:
updates in tests

philbucher committed:
fixing minor syntax errors

philbucher committed:
fixes to make co-sim orchestrated coupling work again

philbucher committed:
all cpp-dummy-solver tests working

philbucher committed:
adding print for size

philbucher committed:
fix c interface to make fortran work

philbucher committed:
update fortran solver

philbucher committed:
making some fortran tests work

philbucher committed:
adding todo comments

philbucher commented:
alright this is ready for review now :)

There are still some minor things left to do but the backbone is done.

The big new thing is that we can work with the same Data across 3 languages (fortran => C => C++ & other way) WITHOUT copying data to a different data structure when passing inetween the languages!

In addition we can use the native data-structures of each language to interact with the IO (except fortran, because from what I read memory should be alloc/dealloc in the same language. This however is only a different fct-call in the end, nothing hackish or inconvenient)

=> In a nutshell this means that the external solver will copy data from its internal data-structure into sth that is being passed to the IO. Inside the IO I directly operate with this data, without duplicating it again. This is basically achieved by using wrappers around the incoming containers. This way I even achieved "resizing" with raw pointers ;) which makes it very easy to work with. In the end inside the IO one can basically use same the functionalities provided by std::vector

I added many tests, but ofc this has still to be cleaned up further
Also to mention that I transformed one of the existing tests (Mok-FSI) and now run with the same setup, but the structure runs in a different mem-space and uses the IO to Communicate => ofc same results ;)

@pooyan-dadvand the details are quite technical, probably we can speak at some point.

philbucher committed:
removed leftover print

philbucher committed:
enabling in Appveyor - ONLY TO SEE IF IT COMPILES IN WIN

philbucher committed:
disabling helpers since require fortran compiler

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher committed:
adding header

philbucher committed:
improving descriptions

philbucher committed:
renamed folder

philbucher committed:
renamed folder

philbucher committed:
mior update general redme

philbucher committed:
improving readmes

philbucher committed:
leftover

philbucher committed:
minor

philbucher committed:
separated file

philbucher committed:
minor cleanup

philbucher committed:
fixed typo

philbucher committed:
Merge remote-tracking branch 'origin/cosim/cosim_io' into cosim/cosim_io

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher committed:
adding github ref

philbucher committed:
minor update readme

philbucher commented on applications/CoSimulationApplication/custom_python/add_co_sim_io_to_python.cpp:
```diff
@@ -0,0 +1,478 @@
+#define KRATOS_CO_SIM_ERROR KRATOS_ERROR
+#define KRATOS_CO_SIM_ERROR_IF KRATOS_ERROR_IF
+#define KRATOS_CO_SIM_ERROR_IF_NOT KRATOS_ERROR_IF_NOT
+#include "co_sim_io/co_sim_io.h"
```
Macros part I

@pooyan-dadvand we already discussed the macro-issue, I think I solved it now, see here

pooyan-dadvand replied:
I'm not sure if I got it correctly. What happend if those macros are not defined?

philbucher replied:
see this file:
https://github.com/KratosMultiphysics/Kratos/blob/cosim/cosim_io/applications/CoSimulationApplication/co_sim_io/impl/co_sim_io_macros.h

If the macros are not predefined (which I am doing here, since we directly want to use the Kratos macros) then it uses its own
I don't know TBH though how this could work for other codes. I basically did what you told me (or at least what I understood)

philbucher commented on applications/CoSimulationApplication/co_sim_io/impl/co_sim_io_macros.h:
```diff
@@ -0,0 +1,48 @@
+//
+
+#ifndef KRATOS_CO_SIM_IO_MACROS_H_INCLUDED
+#define KRATOS_CO_SIM_IO_MACROS_H_INCLUDED
```
Marcos part II

philbucher commented on kratos/input_output/logger.h:
```diff
@@ -245,59 +245,59 @@ namespace Kratos
 #define KRATOS_CHECK_POINT(label) \
   if (false)                      \
-    Logger(label) << Logger::Category::CHECKING
+    Kratos::Logger(label) << Kratos::Logger::Category::CHECKING
```
Macros part III

Would this change be ok for you? I think it wouldn't hurt, it is needed such that I can use the macros outside the `Kratos` namespace

philbucher reviewed (commented):
@pooyan-dadvand I added comments to what I think is the biggest open question

also there are quite a few things left as TODOs, but I am starting to loose the overview. Hence my suggestion would be to take a look at the Macros-issue and then merge the rest more or less as is, then i can address the remaining things properly one by one
Overall the implementation I would say is fine

@adityaghantasala what is your opinion here?

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/c/co_sim_c_io.cpp:
```diff
@@ -0,0 +1,144 @@
+
+void CoSimIO_ImportData(
+    const char* pConnectionName,
+    const char* pIdentifier,
```
`pDataIdentifier`?

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/c/co_sim_c_io.cpp:
```diff
@@ -0,0 +1,144 @@
+void CoSimIO_ImportData(
+    const char* pConnectionName,
+    const char* pIdentifier,
+    int* pSize,
```
We should somehow distinguish more clearely the input and output arguments

philbucher replied:
usually `const` means input, non-`const` is output

pooyan-dadvand replied:
This is absolutely correct. However I would go for some naming convention as peopole may not be so trained...

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/c/co_sim_c_io.cpp:
```diff
@@ -0,0 +1,144 @@
+    double** ppData)
+{
+    using namespace CoSimIO::Internals;
+    std::unique_ptr<DataContainer<double>> p_container(new DataContainerRawMemory<double>(ppData, *pSize));
```
I had to check the code to see if `DataContainerRawMemory` makes a copy or hold the given buffer. I think namimg can be improved.

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/c/co_sim_c_io.cpp:
```diff
@@ -0,0 +1,144 @@
+    int** ppElementTypes)
+{
+    using namespace CoSimIO::Internals;
+    std::unique_ptr<DataContainer<double>> p_container_coords(new DataContainerRawMemory<double>(ppNodalCoordinates, *pNumberOfNodes));
```
I'm not sure if I understood well. But shall I know the number of nodes and elements which I import?

I think you may need a helper function like GetMeshInfo which gives you such data. What do you think?

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/c/co_sim_c_io.cpp:
```diff
@@ -0,0 +1,144 @@
+{
+    using namespace CoSimIO::Internals;
+    std::unique_ptr<DataContainer<double>> p_container_coords(new DataContainerRawMemory<double>(pNodalCoordinates, NumberOfNodes));
+    std::unique_ptr<DataContainer<int>> p_container_conn(new DataContainerRawMemory<int>(pElementConnectivities, NumberOfElements)); // using "NumberOfElements" here is wrong! => has to be computed! Or sth like this ... (maybe even has to be passed...)
```
Why is wrong? If I am exporting the mesh I know the number of nodes and elements. Isn't it?

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/c/co_sim_c_io.cpp:
```diff
@@ -0,0 +1,144 @@
+void CoSimIO_RegisterDataExchangeFunction(
+    const char* pConnectionName,
+    const char* pFunctionName,
+    void (*pFunctionPointer)(const char*, const char*))
```
The signature shouldn't have the dobule** pdata to be passed and size of it?

philbucher replied:
no, it is correct:
The function that is being registered is the one that knows how to import / export things. Hence it only get the `ConnectionName` and the `Identifier` as input and then inside calls the `IO` functions
see here: 
https://github.com/KratosMultiphysics/Kratos/blob/db0b15ecc453d20925cab59eb85cd765c2ab4fc6/applications/CoSimulationApplication/custom_helpers/dummy_solver/cpp/dummy_solver.cpp#L241

and here:
https://github.com/KratosMultiphysics/Kratos/blob/db0b15ecc453d20925cab59eb85cd765c2ab4fc6/applications/CoSimulationApplication/custom_helpers/dummy_solver/cpp/dummy_solver.cpp#L137-L146

I.e. from CoSimulation I would call this function with e.g. "FluidSolver" as `ConnectionName` and "SurfaceTractions" as `Identifier`. The "FluidSolver" internally knows how to get the "SurfaceTractions" from its datastructure, and then calls `CoSimIO` using this data

pooyan-dadvand replied:
Now I see, so the idea is to avoid the dependency on the type of data. Isn't it?

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/c/co_sim_c_io.cpp:
```diff
@@ -0,0 +1,144 @@
+    CoSimIO::IsConverged(pConnectionName, *pConvergenceSignal);
+}
+
+void _AllocateMemoryInt(const int* pSize, int** ppData)
```
Where this would be used? Do solvers need to use these methods to allocate memory for data buffer?

philbucher replied:
It is actually described in the header
=> it is only to be used from Fortran
Reason is that I read (didn't test myself but scared of it :sweat_smile:) one should allocate memory in the same language
I think theoretically one could also allocate the memory inside of fortran, but then we are mixing it and I have no idea what would happen...
We can discuss it though

pooyan-dadvand replied:
Hmmm, I think we should reiterate this. AFAIK, It is important to allocate and delete in the same language but not using it.

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/co_sim_io.h:
```diff
@@ -0,0 +1,96 @@
+
+inline void IsConverged(const std::string& rConnectionName, int& rConvergenceSignal);
+
+inline void SendControlSignal(
```
ReceiveControlSignal wouldn't be needed?

philbucher replied:
Well this is kinda the special thing about CoSimulation
This is only used for the "full" coupling, i.e. when we do the function-registrations etc
I.e. CoSimulation can only send those signals, but is not designed to receive them
Receiving them is an implementational detail, [hidden inside the `Run` function](https://github.com/KratosMultiphysics/Kratos/blob/cosim/cosim_io/applications/CoSimulationApplication/co_sim_io/impl/co_sim_connection.h#L145)
I understand that this is confusing, I will try to address this in my written explanation

pooyan-dadvand replied:
Now I see. Indeed I don't have a use case for it.

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/impl/co_sim_communication.h:
```diff
@@ -0,0 +1,190 @@
+private:
+    std::string mConnectionName;
+    int mEchoLevel = 1;
+    bool mIsConnectionMaster = false;
```
I understand that you don't have kratos flags here. But you may just have flags in raw int format and an internal helper function `Is` which tells you if is true or false a certain position. I would be more extensible in the future.

philbucher replied:
ok sounds good

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/impl/co_sim_connection.h:
```diff
@@ -0,0 +1,281 @@
+        }
+    }
+
+    void IsConverged(int& rConvergenceSignal)
```
Wouldn't be better to return the signal instead of passing it?

Meanwhile, do you think that we should add a label to this method to indicate which criteria is converged?

philbucher replied:
I think I had a reason for it but I don't remember

I am not sure if this is necessary, it basically tells the solver in a strongly coupled simulation if it should go to the next timestep or not
see here: 
https://github.com/KratosMultiphysics/Kratos/blob/db0b15ecc453d20925cab59eb85cd765c2ab4fc6/applications/CoSimulationApplication/custom_helpers/dummy_solver/cpp/dummy_solver.cpp#L214-L220

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/impl/co_sim_connection.h:
```diff
@@ -0,0 +1,281 @@
+
+    void (*mpAdvInTime)(double*) = nullptr;
+    void (*mpInitSolStep)()      = nullptr;
+    void (*mpSolSolStep)()       = nullptr;
```
Is it Solve?

philbucher replied:
well, in Kratos-teminology it is `SolveSolutionStep`
=> it is the only function that can be called multiple times within a timestep
I can rename to `Solve` if you want

pooyan-dadvand replied:
I think `SolveSolutionStep` would be better.

philbucher replied:
solved better now

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/impl/co_sim_io_impl.h:
```diff
@@ -0,0 +1,196 @@
+    KRATOS_CO_SIM_ERROR_IF_NOT(HasIO(rConnectionName)) << "Trying to disconnect connection \"" << rConnectionName << "\" which does not exist!" << std::endl;
+
+    GetConnection(rConnectionName).Disconnect();
+    s_co_sim_connections.erase(std::string(rConnectionName));
```
Connection name isn't a string?

philbucher replied:
yes, I think it is a leftover from when I used `char *`

philbucher replied:
was a leftover

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/impl/co_sim_io_internals.h:
```diff
@@ -0,0 +1,141 @@
+    SettingsType settings;
+    std::string current_line;
+    while (std::getline(settings_file, current_line)) {
+        // TODO implement this
```
:construction:

pooyan-dadvand commented on applications/CoSimulationApplication/co_sim_io/impl/co_sim_mpi_communication.h:
```diff
@@ -0,0 +1,44 @@
+       /*
+        Note to self:
+        If I directly use the buffer of the sender, then I have to ensure it can be reused when returning from the sending function. This can be achieved with two variants:
+        - Use blocking communication. This might block the sender for some time
```
For the first implementaion I would go for the blocking communication...

philbucher replied:
actually I am no longer sure if non-blocking is feasible from a technical point of view and if it even makes much sense, so I will definitely first go for blocking

pooyan-dadvand reviewed (commented):
In general, I like the way it evolves. However, sometimes I'm not sure if I am getting the global overview correctly, especially when it comes to the relation between communication and connection classes. Meanwhile, there are some general comments apart from the ones in the code:
1. I think that we may improve or clarify the memory management of the buffers. 
2. We should think about parallelization and resource management in the future. 
3. How sub-stepping works? How we can change the time-step/sub-step from each solver?

philbucher commented:
> In general, I like the way it evolves. However, sometimes I'm not sure if I am getting the global overview correctly, especially when it comes to the relation between communication and connection classes. Meanwhile, there are some general comments apart from the ones in the code:

I will address this in a separate Doc with UML, I think this will clearify it


> I think that we may improve or clarify the memory management of the buffers.

I agree, this is not trivial. I will also add it in the Doc

> We should think about parallelization and resource management in the future.

What exactly do you mean?

> How sub-stepping works? How we can change the time-step/sub-step from each solver?

Exactly the same as inside of Kratos, mainly by using `AdvanceInTime`. It is however still not clear how the implementation would look in detail. If you want we could discuss it

pooyan-dadvand commented:
> > We should think about parallelization and resource management in the future.
> 
> What exactly do you mean?

What if we have 48 MPI processes how we want to distribute each solver partition instead of having all of them in all processes. 

> 
> > How sub-stepping works? How we can change the time-step/sub-step from each solver?
> 
> Exactly the same as inside of Kratos, mainly by using `AdvanceInTime`. It is however still not clear how the implementation would look in detail. If you want we could discuss it

It would be great to discuss it and give it another iteration.

philbucher commented:
> What if we have 48 MPI processes how we want to distribute each solver partition instead of having all of them in all processes.

this is up to the user, depends on the application-case
For FSI the fluid would typically get more ...

I will include it in my description

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher committed:
missing include

philbucher committed:
started with python exposure of CoSimIO

philbucher committed:
major simplification of installing

philbucher committed:
clean and fix CoSimApp CMakeLists

philbucher committed:
cleanup and fix CMakeLists for CoSimIO

philbucher committed:
adding readme and gitignore

philbucher committed:
standalone py-version working

philbucher committed:
cleanup

philbucher committed:
final changes

philbucher committed:
adding explicit

philbucher committed:
WIP simplifying function passing

philbucher committed:
missing include

philbucher committed:
adding some minor TODOs

philbucher committed:
more work in py-interface

philbucher committed:
added missing types

philbucher committed:
WIP using new Implementation

philbucher committed:
correct install path

philbucher committed:
WIP II using new features

philbucher committed:
temp adding script to show idea

philbucher commented:
I refactored many things related to the function-registration
=> this was overdue (it was a bit a mess) and now made the addition of the Python-interface straight-forward

philbucher commented on applications/CoSimulationApplication/co_sim_io/python/test_import.py:
```diff
@@ -0,0 +1,38 @@
+# CoSimIO.Run(connection_name)
+
+
+CoSimIO.Disconnect(connection_name)
```
@inigolcanalejo this is veryvery roughly how you would use it, I will push more in the next days.
The necessary changes in the backend are done, with this the things you need are really minimal
I will add a more elaborate example soon

philbucher commented:
BTW @pooyan-dadvand @adityaghantasala while using it I realized again how nice/very convenient it is that we made it header-only and independent of Kratos!

philbucher committed:
typo

adityaghantasala commented:
@philbucher  I told you ... ! :-p

philbucher committed:
IsConverged returns now

philbucher committed:
using new Run

philbucher committed:
completing py-interface

philbucher committed:
adding Inigos TAU solver

philbucher committed:
added link to pybind-docs

philbucher committed:
adding TODO

philbucher committed:
adding TAUSolver & Wrapper

WIP

philbucher committed:
adding pybind numpy header
Now it should be possible to pass directly numpy-arrays

philbucher committed:
passing settings as dict works

philbucher committed:
initial support for Communication of files in folder
Requires C++17 if one wants to use it

philbucher committed:
leftover string conversion

philbucher commented:
@pooyan-dadvand @adityaghantasala I am thinking how to design the Interface when running in MPI
THIS WILL NOT BE INCLUDED IN THIS PR
Nevertheless I would like to have at least thought abt this in case it imposes some restrictions or whatsoever.
My current idea is that instead of calling `Connect`, the external solver would call `Connect_MPI` and pass an MPI-Comm that contains the ranks that are connected. 
E.g. for FSI the fluid would need to call `Connect_MPI` on the interface-ranks
This should be the only modification necessary for the external solver, the rest I can handle internally
What do you think? I am very open for suggestions, this is just a proposal:

![image](https://user-images.githubusercontent.com/25484702/79283119-27332680-7eb7-11ea-9080-bd6eef50b533.png)

philbucher commented:
@pooyan-dadvand I thought you made a comment somewhere saying that in addition to the data and the identifier, the interface functions should also get sth like the Kratos-Flags to be able to pass more options
i.e. instead of the current solution:
![image](https://user-images.githubusercontent.com/25484702/79283395-f99aad00-7eb7-11ea-8743-3f336540bbbf.png)

we would have sth like this:
![image](https://user-images.githubusercontent.com/25484702/79283469-3d8db200-7eb8-11ea-9480-f2470a3eb27c.png)

(same probably for the Im/ExportMesh & Im/ExportGeometry)

=> I looked again through the MPCCI docs where I thought I had seen sth similar, but I couldn't find it and I am not even sure if I saw it there to begin with.

Did we discuss this at some point?

I am not sure if it is necessary. It would make the implementation more complicated
I think the `Identifier` would be enough but maybe you have sth different in mind

@adityaghantasala what do you think?

philbucher committed:
updating test-script

philbucher committed:
Update solver file

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

philbucher committed:
removing SendControlSignal from Interface

philbucher committed:
Merge remote-tracking branch 'origin/master' into cosim/cosim_io

pooyan-dadvand commented:
> @pooyan-dadvand I thought you made a comment somewhere saying that in addition to the data and the identifier, the interface functions should also get sth like the Kratos-Flags to be able to pass more options
> i.e. instead of the current solution:
> ![image](https://user-images.githubusercontent.com/25484702/79283395-f99aad00-7eb7-11ea-8743-3f336540bbbf.png)
> 
> we would have sth like this:
> ![image](https://user-images.githubusercontent.com/25484702/79283469-3d8db200-7eb8-11ea-9480-f2470a3eb27c.png)
> 
> (same probably for the Im/ExportMesh & Im/ExportGeometry)
> 
> => I looked again through the MPCCI docs where I thought I had seen sth similar, but I couldn't find it and I am not even sure if I saw it there to begin with.
> 
> Did we discuss this at some point?
> 
> I am not sure if it is necessary. It would make the implementation more complicated
> I think the `Identifier` would be enough but maybe you have sth different in mind
> 
> @adityaghantasala what do you think?

If I remember correctly, my idea was to use a structure/class for this extra parameter to be extensible for for future addition of new parameters.
